src/main/resources/category/java/concurrent.xml

```diff
@@ -169,6 +169,1424 @@ class Exporter {
+            &#13;5. Use package private and @VisibleForTesting for methods used for JUnit only.
+            (jpinpoint-rules)</description>
+        <priority>1</priority>
+        <properties>
+            <property name="tags" value="jpinpoint-rule,multi-threading,data-mix-up" type="String" description="classification"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+(: known assumed singleton types by annotation :)
+//ClassDeclaration[
+   (
+       pmd-java:hasAnnotation('org.springframework.stereotype.Service')
+    or pmd-java:hasAnnotation('org.springframework.stereotype.Component')
+    or pmd-java:hasAnnotation('org.springframework.stereotype.Controller')
+    or pmd-java:hasAnnotation('org.springframework.stereotype.Repository')
+    or pmd-java:hasAnnotation('org.springframework.web.bind.annotation.RestController')
+    or (pmd-java:hasAnnotation('javax.ejb.Singleton') and ModifierList/Annotation[pmd-java:typeIs('javax.ejb.ConcurrencyManagement')]/AnnotationMemberList/MemberValuePair/FieldAccess[./TypeExpression[pmd-java:typeIs('javax.ejb.ConcurrencyManagementType')] and @Name="BEAN"])
+    or (pmd-java:hasAnnotation('jakarta.ejb.Singleton') and ModifierList/Annotation[pmd-java:typeIs('jakarta.ejb.ConcurrencyManagement')]/AnnotationMemberList/MemberValuePair/FieldAccess[./TypeExpression[pmd-java:typeIs('jakarta.ejb.ConcurrencyManagementType')] and @Name="BEAN"])
+   )
+   (: not shared when request or prototype scope :)
+   and not(pmd-java:hasAnnotation('org.springframework.context.annotation.Scope') and .//AnnotationMemberList/MemberValuePair[
+       StringLiteral[contains(@Image,'request') or contains(@Image,'prototype')]
+       or FieldAccess[@Name='SCOPE_REQUEST' or @Name='SCOPE_PROTOTYPE']/TypeExpression[pmd-java:typeIs('org.springframework.beans.factory.config.ConfigurableBeanFactory')]]
+    )
+   and not(pmd-java:hasAnnotation('org.springframework.context.annotation.RequestScope'))
+   (: if @NotThreadSafe no checking :)
+   and not(pmd-java:hasAnnotation('net.jcip.annotations.NotThreadSafe'))
+   (: ConfigurationProperties assumed executed only once, no violation, except if Lombok Setter :)
+   and not(pmd-java:hasAnnotation('org.springframework.boot.context.properties.ConfigurationProperties'))
+   and not(pmd-java:hasAnnotation('lombok.Setter'))]
+(: any method call:)
+//MethodCall[
+    (: class extends a base class :)
+    ancestor::ClassDeclaration//ExtendsList
+    (: not in static inner class :)
+    and not (ancestor::ClassDeclaration[@Nested=true() and @Static=true()])
+    (:  has arguments (a method call), and  a compound like date.setTime, map.put, list.set :)
+    and matches(@MethodName,"(set|put|clear|add|remove|replace)")
+    (: not a field of this class, no param or local :)
+    and ./AmbiguousName
```

Comment:
hoezo deze, AmbigousName? Dit is als het type niet ge-resolved kan worden toch?

Reply:
Ah yes, interesting case. Below the `date` is AmbigousName, because the `SynchronizedTryout` class is not found (or on classpath?). 

```java
@Component
public class MutateInheritedFieldTryout extends SynchronizedTryout {
    final Map mapField = getMap(); // found by other rule
    private String removed; // found by other rule

    public String getDateString() {
        return date.toString();
    }
...
```

Now if I add the following in the same file:

```java
class SynchronizedTryout {
    protected Date date;
}
```

It will resolve to `VariableAccess` with a `@Name` and Type.

Can we assume that the base class is always available when checking the code? Than I can replace the AmbigousName with an actual check.

Reply:
I think so, try out with a unit test

Reply:
Yep, I will, but the `SynchronizedTryout` is a class in the same file, so sure it will find it in unit test... Maybe I'll try to extend an know java class as well and see if that works (if I can find one with exposed vars in base class).

Reply:
made fix